### PR TITLE
Fix #3205: [Rtl] High-fi Home Screen All Topics List

### LINKED ISSUE #3205: [RTL] High-fi Home Screen All Topics List
All Topics List items needs to support RTL
- Spacing between the items needs to be fixed
- Texts inside the card needs to be right aligned and reversed

Mobile Portrait: https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/eabb7da8-e5e5-418d-b3d9-d59ae0b5f585/specs/
Mobile Landscape: https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/0dbdd000-6738-4f6b-9026-0d16737b7baa/specs/
Tablet : Landscape https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/screen/5434c52d-b32b-4666-8b28-cf03b3cbd4cd/
Portrait: https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/screen/b3019177-fe86-459e-8a63-8109c53005f5/specs/

### PR DESCRIPTION
**Pull request by veena14cs**
## Explanation
This PR fixes #3205. 

## Screenshot LTR and RTL

screenshot  .... screenshot

screenshot  ...  screenshot
`1324e566tr6

Tablet Portrait: 
screenshot ....  screenshot

Landscape:

screenshot........screenshot

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by veena14cs** [comment]
@rt4914 Can you please look into this PR. There is issue with Tablet landscape mode. I am not getting equal spacing around the items. I tried setting margins in different ways but its not working.

**Comment by veena14cs** [comment]
screenshot

**Comment by rt4914** [comment]
@veena14cs I haven't been able to solve this today remained stuck on it. Will try again tomorrow with fresh mind.

**Comment by veena14cs** [comment]
> @veena14cs I haven't been able to solve this today remained stuck on it. Will try again tomorrow with fresh mind.

Ok @rt4914 did you find any fix?

**Comment by rt4914** [comment]
> > @veena14cs I haven't been able to solve this today remained stuck on it. Will try again tomorrow with fresh mind.
> 
> Ok @rt4914 did you find any fix?

@veena14cs Haven't been able to find any fix. Also I am unable to spend a lot of time on this so you might need to try some different methods to find the problem and then solve it.

**Comment by veena14cs** [comment]
> > > @veena14cs I haven't been able to solve this today remained stuck on it. Will try again tomorrow with fresh mind.
> > 
> > 
> > Ok @rt4914 did you find any fix?
> 
> @veena14cs Haven't been able to find any fix. Also I am unable to spend a lot of time on this so you might need to try some different methods to find the problem and then solve it.

Thanks @rt4914 I will try out different solutions. 
@BenHenning do you have any suggestions on this.

**Comment by BenHenning** [comment]
@veena14cs I'm probably not going to have bandwidth to look into this this week. You may need to step debug into the actual UI components that we're using to better understand the spacing logic and why the observed behavior is happening in order to figure out how to fix it (this is what I would do if I were to dig into this problem in detail). It's difficult for us to help without more context, so this seems to be the next step to take.

**Comment by veena14cs** [comment]
> @veena14cs I'm probably not going to have bandwidth to look into this this week. You may need to step debug into the actual UI components that we're using to better understand the spacing logic and why the observed behavior is happening in order to figure out how to fix it (this is what I would do if I were to dig into this problem in detail). It's difficult for us to help without more context, so this seems to be the next step to take.

Thanks Ben! I have fixed the issue, it will be good if you can please review this PR. I have added tablet landscape mode screenshot in PR description for reference. The issue was there is no equal spacing for the items. Solution was to add equal spacing for all the items.

@rt4914 PTAL

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import android.view.View
+import androidx.core.view.ViewCompat
+import androidx.recyclerview.widget.GridLayoutManager
+import androidx.recyclerview.widget.OrientationHelper
+import androidx.recyclerview.widget.RecyclerView
+import androidx.recyclerview.widget.RecyclerView.ItemDecoration
+import org.oppia.android.R
+import org.oppia.android.app.home.HomeFragmentPresenter
+
+class GridSpacingItemDecoration(context: Context) : ItemDecoration() {
```
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:20]: Add Kdoc for this file.

File: app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt [hunk 2]
```diff
@@ -77,6 +78,8 @@ class HomeFragmentPresenter @Inject constructor(
         else spanCount
       }
     }
+    binding.homeRecyclerView.addItemDecoration(GridSpacingItemDecoration(activity));
```
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt]: Remove `;` at the end.

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    val childIndex = parent.getChildAdapterPosition(view)
+    val itemSpanSize = getItemSpanSize(parent, childIndex)
+    val spanIndex = getItemSpanIndex(parent, childIndex)
+
+    val position = parent.getChildAdapterPosition(view)
+    val viewType = parent.adapter!!.getItemViewType(position)
+
+    if (viewType == HomeFragmentPresenter.ViewType.TOPIC_LIST.ordinal) {
+//
+      /* INVALID SPAN */if (spanCount < 1) return
```
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt]: Remove this comment and above `//` too.

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+      if (spanCount == 2) {
+        outRect.right = halfSpacing
+        outRect.left = halfSpacing
+      }
+
+      if (spanCount == 4) {
+        outRect.left =
+          (spacing ) - spanIndex * (spacing ) / spanCount // spacing - column * ((1f / spanCount) * spacing)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt]: Nit:
`(spacing) - spanIndex * (spacing) / spanCount // spacing - column * ((1f / spanCount) * spacing)`

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        }
+      }
+
+      if (isLayoutRTL(parent)) {
+        val tmp = outRect.left
+        outRect.left = outRect.right
+        outRect.right = tmp
+      }
+
+      if (spanIndex < spanCount) { // top edge
```
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt]: What does `top edge` means here?

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  private fun getTotalSpan(parent: RecyclerView): Int {
+    val mgr = parent.layoutManager
+    if (mgr is GridLayoutManager) {
+      return mgr.spanCount
+    }
+    return -1
+  }
+
+  private fun getItemSpanSize(parent: RecyclerView, childIndex: Int): Int {
+    val mgr = parent.layoutManager
```
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt]: What is `mgr`? Can we give a better name for clear understanding? to me it looks like `layoutManager`

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt (hunk 1 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:20]: Also please add reference from where this code has been used.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@veena14cs Suggested changes.

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt (hunk 1 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:20]: Done.

File: app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt (hunk 2 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/home/HomeFragmentPresenter.kt]: Done.

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt (hunk 3 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt]: Done.

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt (hunk 4 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt]: Done.

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt (hunk 5 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt]: Top margin.

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt (hunk 6 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt]: Done.

**Comment by veena14cs** [comment]
Thanks @rt4914 fixed changes PTAL.

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt (hunk 1 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:20]: Nit
```
/** 
 * An [RecyclerView.ItemDecoration] that adds spacing between cells in a [GridLayoutManager].
 * 
 * Reference: https://stackoverflow.com/a/30701422/4500099
 */
```

**Review by rt4914 — CHANGES_REQUESTED** [review]
@BenHenning Any idea how to approach towards adding test for this new `GridSpacingItemDecoration` file. I think its more of `Screenshot` testing part but still if you have any other suggestions.

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt (hunk 1 above)
Comment (rt4914) [reply app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:20]: @veena14cs PTAL

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+      return gridLayoutManager.orientation
+    }
+    return VERTICAL
+  }
+
+  companion object {
+    private const val VERTICAL = OrientationHelper.VERTICAL
+
+    @SuppressLint("NewApi", "WrongConstant")
+    private fun isLayoutRTL(parent: RecyclerView): Boolean {
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt]: Nit: ``isRtlLayout`` (switched the word order, and we should say 'Rtl' not 'RTL' per camel-casing guidelines in the style guide).

**Review by BenHenning — COMMENTED** [review]
@rt4914 @veena14cs could we maybe test this with Robolectric? If we lay out a test FrameLayout that contains the RecyclerView with the decorator attached, we should be able to create different situations to test the decoration.

I agree that Espresso + screenshot tests would probably be easier, but likely not thorough. Screenshot tests should verify visual correctness, not functional correctness. We should still try to use unit-level tests for that.

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt (hunk 7 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt]: Thanks! done.

File: app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt (hunk 1 above)
Comment (veena14cs) [reply app/src/main/java/org/oppia/android/app/recyclerview/GridSpacingItemDecoration.kt:20]: Done.

**Comment by rt4914** [comment]
> @rt4914 @veena14cs could we maybe test this with Robolectric? If we lay out a test FrameLayout that contains the RecyclerView with the decorator attached, we should be able to create different situations to test the decoration.
> 
> I agree that Espresso + screenshot tests would probably be easier, but likely not thorough. Screenshot tests should verify visual correctness, not functional correctness. We should still try to use unit-level tests for that.

@veena14cs One of the ways we can test this is by using an item and this decorator such that if there is (decorator + item) then we can display 3 items in grid but if its only item without decorator then the grid shows 4 or more items which automatically proves that the decorator is taking taking up some space. We can use this for different qualifiers and check.

@veena14cs WDYT?

**Comment by veena14cs** [comment]
> > @rt4914 @veena14cs could we maybe test this with Robolectric? If we lay out a test FrameLayout that contains the RecyclerView with the decorator attached, we should be able to create different situations to test the decoration.
> > I agree that Espresso + screenshot tests would probably be easier, but likely not thorough. Screenshot tests should verify visual correctness, not functional correctness. We should still try to use unit-level tests for that.
> 
> @veena14cs One of the ways we can test this is by using an item and this decorator such that if there is (decorator + item) then we can display 3 items in grid but if its only item without decorator then the grid shows 4 or more items which automatically proves that the decorator is taking taking up some space. We can use this for different qualifiers and check.
> 
> @veena14cs WDYT?

Just for clarification, should I test span count over here? There is already Roboelectric test `HomeSpanTest` and it passes all the checks.

**Comment by veena14cs** [comment]
This issue is getting fixed in #3452 . Once that is verified and merged will be closing this PR.

**Comment by veena14cs** [comment]
This issue is fixed in #3452. Therefore, closing this PR.

**Pull request closed without merging**